### ISSUE TITLE
Wrong NUGET_PACKAGES folder

### ISSUE THREAD

**Issue by kmadof** [MicrosoftDocs/azure-devops-docs#7572]
In NuGet package caching section this folder is given:

> $(Pipeline.Workspace)/.nuget/packages

However for me it was not working. I checked it on Linux and Windows host agents and this folder simply doesn't exist. (I was checking it after restoring NuGet packages). However dotnet cli was handy here:

> dotnet nuget locals global-packages -l

gave me for:

- Linux - info : global-packages: /home/vsts/.nuget/packages/
- Windows - info : global-packages: D:\home\vsts\.nuget\packages

Did I miss something like setting global-packages in Nuget.config?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 77c20289-dab2-7a6e-4f63-4b30eaff2c71
* Version Independent ID: d75cc196-90c1-8291-9907-ff405bde4edf
* Content: [Pipeline caching - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#netnuget)
* Content Source: [docs/pipelines/release/caching.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/release/caching.md)
* Product: **devops**
* Technology: **devops-cicd-release**
* GitHub Login: @ramiMSFT
* Microsoft Alias: **rabououn**

**Comment by WilliamAntonRohm** [comment]
@kmadof -- Krzysztof, thank you for your question.  You may find answers here:
- [Azure DevOps Support Bot](https://azuredevopsvirtualagent.azurewebsites.net/) 
- [Azure DevOps on Stack Overflow](https://stackoverflow.com/questions/tagged/azure-devops) 

@ramiMSFT -- Rami, please look into this issue.

**Comment by kmadof** [comment]
@WilliamAntonRohm I solved that using this step:

```
- task: PowerShell@2
  inputs:
    targetType: 'inline'
    script: |
      $cache = dotnet nuget locals global-packages -l
      $cacheLocation = $cache.Replace('info : global-packages: ', '')
      
      Write-Host $cacheLocation
      
      Write-Host "##vso[task.setvariable variable=NUGET_PACKAGES;]$cacheLocation"

    pwsh: true
```

I just simply wonder if I did something wrong or maybe documentation needs to be updated. I spent few hours to solve it so maybe there is a chance to save others people effort, by maaking documentation more precise.

**Comment by ramiMSFT** [comment]
@Kmdof Thank you for your contribution. Before caching your NuGet package you have to properly configure a cache task. Please refer to this [link](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#configuring-the-task) for guidance.
Also please keep in mind that on the first run after the task is added, the cache step will report a "cache miss" since the cache identified by that key does not exist. After the last step, a cache will be created from the files in $(Pipeline.Workspace)/.nuget and uploaded. On the next run, the cache step will report a "cache hit" and the contents of the cache will be downloaded and restored.
Thank you for your suggestion. I'm glad your issue is resolved.

**Comment by kmadof** [comment]
@ramiMSFT I had task configured as it is described [here](https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#configuring-the-task). However, variable $(Pipeline.Workspace) used there was evaluated to path `D:\a\1` and there is no `.nuget/packages`, but I found packages restored in ` D:\home\vsts.nuget\packages` and I cached this folder. It works fine, but I'm not sure it this is a mistake in doc?

To sum up, path given in doc where we suppose find nuget packages 
```
variables:
  NUGET_PACKAGES: $(Pipeline.Workspace)/.nuget/packages
```
doesn't exist. 

And yes it was like you described, in first run 'cache miss' then 'cache hit', but not with $(Pipeline.Workspace)/.nuget 

![image](https://user-images.githubusercontent.com/1680172/76766692-ab9b6800-6798-11ea-9681-9cbff7347ca6.png)

**Comment by jenshnielsen** [comment]
@kmadof I have not tried this my self but according to https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-environment-variables setting NUGET_PACKAGES should change the dir where the packages are located

**Comment by rafek1241** [comment]
@ramiMSFT 
Why did you close it? In my opinion it could be helpful for people like me that are struggling with the same issue.

**Comment by rafek1241** [comment]
@WilliamAntonRohm Thank you!

**Comment by WilliamAntonRohm** [comment]
@rafek1241 -- Robert, thank you for your follow-up!

@ramiMSFT -- Rami, please look into this potential documentation issue.

**Comment by msmolka** [comment]
I'm also struggling with this. Cache is restored but then `dotnet build` is not seeing restored packages from. And I believe this is the issue related with this variable.

**Comment by Nirmal4G** [comment]
I did hit this too. Instead of setting `NUGET_PACKAGES`, set `NuGet.Packages` which will automatically create `NUGET_PACKAGES` environment variable properly. Hope this helps.

@WilliamAntonRohm @ramiMSFT Please update the docs if this solution is okay with you!

**Comment by bjarte** [comment]
@Nirmal4G Can you elaborate your solution?

Did you simply change these lines in the documentation
```
variables:
  NUGET_PACKAGES: $(Pipeline.Workspace)/.nuget/packages
```

to this?
```
variables:
  NuGet.Packages: $(Pipeline.Workspace)/.nuget/packages
```

**Comment by bjarte** [comment]
When I follow the documentation to the letter and then run nuget restore after successfully having cached the nuget directory, I get a lot of these from the nuget restore step:
Added package 'Castle.Core.4.4.0' to folder 'D:\a\1\s\packages' from source 'D:\a\1\.nuget\packages'

So it seems the nugets are not stored in the correct directory and the nuget restore task has to copy/move all of them over.

**Comment by Nirmal4G** [comment]
@bjarte Yes, that's correct. Azure pipelines will convert Variables of the form `My.Env.Var` to `MY_ENV_VAR`. It no longer respects setting the Env var directly.

**Comment by bjarte** [comment]
I still haven't managed to skip the 'Nuget restore' task completely (which the documentation says I should be able to), but I found this workaround to make the process quicker:

```
variables:
  NUGET_PACKAGES: $(System.DefaultWorkingDirectory)/packages`

  - task: CmdLine@2
    displayName: Create packages directory
    inputs:
      script: |
        mkdir packages
```
This is the same directory that the Nuget restore step uses.

Then when I run the 'Nuget Restore', all it does is this for each package:
`Checking compatibility for System.Threading.AccessControl 4.6.0 with .NETFramework,Version=v4.8.`

So the process is pretty quick, it doesn't have to move or downloading anything.

**Comment by maks200179** [comment]
My question somehow related to this topic.

In my case i found that its correct path to use. 

`- task: Cache@2
  displayName: 'NuGet Cache'
  inputs:
    key: 'nuget | "$(Agent.OS)" | **/packages.lock.json,!**/bin/**,!**/obj/**'
    restoreKeys: |
       nuget | "$(Agent.OS)"
       nuget
    path: '$(System.DefaultWorkingDirectory)\\project_name/packages'
    cacheHitVar: 'CACHE_RESTORED'`

But after the cache restored the path of packages is wrong that why my msbuild task fail. 

For example the correct path is packages\Microsoft.Bcl.Build.1.0.21\build\Microsoft.Bcl.Build.targets.
But after restored from azure devops cache it is under wrong path packages\microsoft.bcl.build\1.0.21\build\Microsoft.Bcl.Build.targets.

**Comment by WilliamAntonRohm** [comment]
Related to #12985.

**Comment by Aeroverra** [comment]
Also having this issue. Thanks to everyone who gave work arounds. Dev ops has prooven to be a time suck because of small documentation errors like this. Please fix.

**Comment by heku** [comment]
Following the official docs leads me to here.

**Comment by craigbrown** [comment]
Also had this issue - but I didn't realize that the `NUGET_PACKAGES` variable that is declared in the example is actually setting an environment variable behind the scenes, which is then used by the nuget restore process. I was trying to use the path directly.

Might be worth pointing out that setting the value of that variable is actually quite important!
